h2o-algos/src/main/java/hex/rulefit/Rule.java

```diff
@@ -0,0 +1,183 @@
+        }
+        return hashCode;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return this.hashCode() == obj.hashCode();
+    }
+    
+    public static Set<Rule> extractRulesFromModel(SharedTreeModel model, int modelId) {
+        Set<Rule> rules = new HashSet<>();
+        final SharedTreeModel.SharedTreeOutput sharedTreeOutput = (SharedTreeModel.SharedTreeOutput) model._output;
+        final int treeClass = getResponseLevelIndex(null, sharedTreeOutput);
+        for (int i = 0; i < ((SharedTreeModel.SharedTreeParameters) model._parms)._ntrees; i++) {
+        SharedTreeSubgraph sharedTreeSubgraph = model.getSharedTreeSubgraph(i, treeClass);
+        rules.addAll(extractRulesFromTree(sharedTreeSubgraph, modelId));
+        }
+        
+        return rules;
+    }
+
+    public static List<Rule> extractRulesListFromModel(SharedTreeModel model, int modelId) {
+        List<Rule> rules = new ArrayList<>();
+        final SharedTreeModel.SharedTreeOutput sharedTreeOutput = (SharedTreeModel.SharedTreeOutput) model._output;
+        final int treeClass = getResponseLevelIndex(null, sharedTreeOutput);
+        for (int i = 0; i < ((SharedTreeModel.SharedTreeParameters) model._parms)._ntrees; i++) {
+            SharedTreeSubgraph sharedTreeSubgraph = model.getSharedTreeSubgraph(i, treeClass);
+            rules.addAll(extractRulesFromTree(sharedTreeSubgraph, modelId));
+        }
+
+        return rules;
+    }
+    
+    public static Set<Rule> extractRulesFromTree(SharedTreeSubgraph tree, int modelId) {
+        Set<Rule> rules = new HashSet<>();
+        List<Condition> conditions = new ArrayList<>();
+        rules = traverseNodes(tree.rootNode, conditions, rules, null, modelId);
+        return rules;
+    }
+    
+    private static Set<Rule> traverseNodes(SharedTreeNode node, List<Condition> conditions, Set<Rule> rules, Condition conditionToAdd, int modelId) {
+        if (conditionToAdd != null)
+            conditions.add(conditionToAdd);
+        
+        // extract this node
+        Condition.Type type;
+        Condition.Operator operator;
+        String[] catThreshold;
+        int[] catThresholdNum;
+        
+        if (node.isLeaf()) {
+            // create Rule to return
+            rules.add(new Rule(conditions.toArray(new Condition[]{}), node.getPredValue(), "M" + modelId + "T" + node.getSubgraphNumber() + "N" + node.getNodeNumber()));
+            return rules;
+        } else {
+            // extract condition leading to the right and recurse there
+            if (node.getDomainValues() == null) {
+                type = Condition.Type.Numerical;
+                operator = Condition.Operator.GreaterThanOrEqual;
+                rules.addAll(traverseNodes(node.getRightChild(), new ArrayList<>(conditions), rules, new Condition(node.getColId(), type, operator,  node.getSplitValue(), null, null, node.getColName(), node.getRightChild().isInclusiveNa()), modelId));
```

Comment:
I think you can move new ArrayList<>(conditions) to the `if (conditionToAdd != null)` block

Reply:
also these lines are too long, I would put the new Condition() call on separate line

Reply:
new Condition() call on separate line here 4e8a040